src/syscheckd/whodata/audit_parse.c

```diff
@@ -230,34 +236,110 @@ void clean_regex() {
+    int is_hex_buffer = 1;
+    int end = 0;
+    int key_length = strlen(key);
+
+    // Find the key
+    for (start = strstr(buffer, key); start != NULL; start = strstr(start + 1, key)) {
+        if (start[key_length] != '=') {
+            continue;
+        }
+        // Check if the there is a field that matches `key` argument.
+        if (start == buffer || *(start - 1) == ' ' || *(start - 1) == '\n') {
+            break;
+        }
+    }
+
+    if (start == NULL) {
+        return NULL;
+    }
+
+    start += key_length + 1;
+
+    if (*start == '"') {
+        is_hex_buffer = 0;
+        start++;
+    }
+
+    // The key can be limited by one of these three characters
+    if (end = strcspn(start, "\n \""), end == 0) {
+        return NULL;
+    }
 
-int filterkey_audit_events(char *buffer) {
-    int i = 0;
-    char logkey1[OS_SIZE_256] = { 0 };
-    char logkey2[OS_SIZE_256] = { 0 };
+    os_calloc(end + 1, sizeof(char), value);
+    strncpy(value, start, end);
 
-    snprintf(logkey1, OS_SIZE_256, "key=\"%s\"", AUDIT_KEY);
-    if (strstr(buffer, logkey1)) {
-        mdebug2(FIM_AUDIT_MATCH_KEY, logkey1);
-        return 1;
+    if (is_hex_buffer) {
+        ascii_value = decode_hex_buffer_2_ascii_buffer(value, end);
+        free(value);
+        return ascii_value;
     }
 
-    snprintf(logkey1, OS_SIZE_256, "key=\"%s\"", AUDIT_HEALTHCHECK_KEY);
-    if (strstr(buffer, logkey1)) {
-        mdebug2(FIM_AUDIT_MATCH_KEY, logkey1);
-        return 3;
+    return value;
+}
+
+/**
+ * @brief Looks if the buffer contains a valid audit key (AUDIT_KEY, AUDIT_HC_KEY or a specified key)
+ *
+ * @param buffer Audit message to look for.
+ * @return Type of key.
+ * @retval FIM_AUDIT_UNKNOWN_KEY if the key is unknown.
+ * @retval FIM_AUDIT_KEY if the key of the event is AUDIT_KEY.
+ * @retval FIM_AUDIT_HC_KEY if the key of the event is AUDIT_HEALTHCHECK_KEY.
+ * @retval FIM_AUDIT_CUSTOM_KEY if the key of the event is configured using audit_key option.
+ */
+STATIC audit_key_type filterkey_audit_events(char *buffer) {
+    char *save_ptr = NULL;
+    char *full_key = NULL;
+    char *key = NULL;
+    audit_key_type retval = FIM_AUDIT_UNKNOWN_KEY;
+    int i;
+
+    // Find the key
+    if (full_key = get_audit_field(buffer, "key"), full_key == NULL) {
+        return retval;
     }
 
-    while (syscheck.audit_key[i]) {
-        snprintf(logkey1, OS_SIZE_256, "key=\"%s\"", syscheck.audit_key[i]);
-        snprintf(logkey2, OS_SIZE_256, "key=%s", syscheck.audit_key[i]);
-        if (strstr(buffer, logkey1) || strstr(buffer, logkey2)) {
-            mdebug2(FIM_AUDIT_MATCH_KEY, logkey1);
-            return 2;
+    for (key = strtok_r(full_key, "\001", &save_ptr); key != NULL; key = strtok_r(NULL, "\001", &save_ptr)) {
+        if (*key == '\0') {
+            continue;
+        }
+
+        if (strcmp(key, AUDIT_KEY) == 0) {
+            mdebug2(FIM_AUDIT_MATCH_KEY, full_key);
+            retval = FIM_AUDIT_KEY;
+            goto end;
```

Comment:
The `goto` approach doesn't really pay off here, freeing `full_key` and returning  the appropriate key type takes up the exact same number of lines, is more efficient and I would argue easier to read.

```suggestion
            free(full_key);
            return FIM_AUDIT_KEY;
```